tests/test_main.py

```diff
@@ -60,3 +60,15 @@ def test_from_pydantic_with_invalid_call():
         @from_pydantic("foo")
         def cli(foo):
             pass
+
+
+def test_disabled_default():
```

Comment:
Not sure what is tested here (besides the fact that `from_pydantic` doesn't crash when `defer_set_default=True`), since the expected behavior is identical to `defer_set_default=False`. I'd be curious to see a test case closer to the real case you're mentioning in the description:

- add `pydantic-settings` to dev dependencies
- define a `BaseSettings`-based object and apply Pydantic to it
- add test case for the case where Pydantic default is read from environment variable, and ensure the right default is used in both `defer_set_default=True` and `defer_set_default=False`

Reply:
I added test_base_settings. The main use case here is that when using BaseSettings, there is an order of precedence that is added if fields are undefined (see https://docs.pydantic.dev/latest/concepts/pydantic_settings/#customise-settings-sources).

The default precedence order is:
- cli (this is the pydantic implementation that we ignore in this context)
- init parameters
- environment variables
- other sources
- default

When click defines the default value or a None is passed into the init then it breaks the precedence order because all the values are considered init params and the precedence order becomes:
- cli (this is the pydantic implementation that we ignore in this context)
- init parameters or default from click
- environment variables
- other sources

This has the effect of making environment variables be ignored when they should be used. It is only an issue with BaseSettings because a BaseModel has no concept of precedence.